crates/turbopack-core/src/chunk/mod.rs

```diff
@@ -362,194 +367,305 @@ pub trait FromChunkableAsset: ChunkItem + Sized + Debug {
-                    queue.push_back(ChunkContentWorkItem::AssetReferences(
-                        chunk_item.references(),
+                    graph_nodes.push(ChunkContentGraphNode::ChunkItem(manifest_loader_item));
+                    graph_nodes.push(ChunkContentGraphNode::AsyncChunkGroup(
+                        ChunkGroupVc::from_asset(manifest_chunk, context.chunking_context),
                     ));
-                    chunk_items.push(chunk_item);
-                }
-                chunks.extend(inner_chunks);
-                async_chunk_groups.extend(inner_chunk_groups);
-
-                // Make sure the chunk doesn't become too large.
-                // This will hurt performance in many aspects.
-                let chunk_items_count = chunk_items.len();
-                if !split
-                    && prev_chunk_items != chunk_items_count
-                    && chunk_items_count > 5000
-                    && prev_chunk_items > 1
-                {
-                    // Chunk is too large, cancel this algorithm and
-                    // restart with splitting from the start
-                    return Ok(None);
+                } else {
+                    return Ok(vec![ChunkContentGraphNode::ExternalAssetReference(
+                        reference,
+                    )]);
                 }
             }
         }
     }
 
+    if !new_processed_assets.is_empty() && new_chunk_items_count > 0 {
+        let mut state = state.lock();
+        state.processed_assets.extend(new_processed_assets);
+        state.chunk_items_count += new_chunk_items_count;
+    }
+
+    Ok(graph_nodes)
+}
+
+async fn chunk_item_to_graph_nodes<I>(
+    context: ChunkContentContext,
+    state: Arc<Mutex<ChunkContentState>>,
+    chunk_item: I,
+) -> Result<impl Iterator<Item = ChunkContentGraphNode<I>>>
+where
+    I: FromChunkableAsset + Eq + std::hash::Hash + Clone,
+{
+    Ok(chunk_item
+        .references()
+        .await?
+        .into_iter()
+        .map(|reference| reference_to_graph_nodes::<I>(context, Arc::clone(&state), *reference))
+        .try_join()
+        .await?
+        .into_iter()
+        .flatten())
+}
+
+/// The maximum number of chunk items that can be in a chunk before we split it
+/// into multiple chunks.
+const MAX_CHUNK_ITEMS_COUNT: usize = 5000;
+
+async fn chunk_content_internal_parallel<I>(
+    chunking_context: ChunkingContextVc,
+    entry: AssetVc,
+    additional_entries: Option<AssetsVc>,
+    split: bool,
+) -> Result<Option<ChunkContentResult<I>>>
+where
+    I: FromChunkableAsset + Eq + std::hash::Hash + Clone,
+{
+    let mut processed_assets: HashSet<(ChunkingType, AssetVc)> = HashSet::default();
+
+    // TODO(alexkirsz) We shouldn't need a clone here.
+    let additional_entries = if let Some(additional_entries) = additional_entries {
+        additional_entries.await?.clone_value().into_iter()
+    } else {
+        vec![].into_iter()
+    };
```

Comment:
It should work this way:

```suggestion
    let additional_entries = if let Some(additional_entries) = additional_entries {
        Some(additional_entries.await?)
    } else {
        None
    };
    let additional_entries = additional_entries.into_iter().map(|list| list.iter()).flatten();
```

The first declarations stores the owned thing. The second can access it by reference.
Also [`Option::into_iter()`](https://doc.rust-lang.org/std/option/enum.Option.html#method.iter) is useful here.

Reply:
Doing this (and the below) results in the following error (which actually gets reported in turbopack-ecmascript--ugh):
```
error: implementation of `Iterator` is not general enough
   --> crates/turbopack-ecmascript/src/chunk/mod.rs:332:1
    |
332 | #[turbo_tasks::function]
    | ^^^^^^^^^^^^^^^^^^^^^^^^ implementation of `Iterator` is not general enough
    |
    = note: `Iterator` would have to be implemented for the type `std::option::Iter<'0, ReadRef<Assets, Vec<AssetVc>>>`, for any lifetime `'0`...
    = note: ...but `Iterator` is actually implemented for the type `std::option::Iter<'1, ReadRef<Assets, Vec<AssetVc>>>`, for some specific lifetime `'1`
    = note: this error originates in the attribute macro `turbo_tasks::function` (in Nightly builds, run with -Z macro-backtrace for more info)

error: implementation of `FnOnce` is not general enough
   --> crates/turbopack-ecmascript/src/chunk/mod.rs:332:1
    |
332 | #[turbo_tasks::function]
    | ^^^^^^^^^^^^^^^^^^^^^^^^ implementation of `FnOnce` is not general enough
    |
    = note: closure with signature `fn(&'0 ReadRef<Assets, Vec<AssetVc>>) -> std::slice::Iter<'_, AssetVc>` must implement `FnOnce<(&ReadRef<Assets, Vec<AssetVc>>,)>`, for any lifetime `'0`...
    = note: ...but it actually implements `FnOnce<(&ReadRef<Assets, Vec<AssetVc>>,)>`
    = note: this error originates in the attribute macro `turbo_tasks::function` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I'll take another look at it later.